File: server/python/app/routes/llm.py
from flask import Blueprint, request, jsonify
from app.models.schemas import LLMModel
from app.utils import db_utils
from app.utils import llm_utils
import urllib.parse

llm_bp = Blueprint("llm", __name__)


@llm_bp.route("/models", methods=["GET"])
def get_models():
    """Get all available LLM models."""
    # Return both default models and custom models
    models = db_utils.get_llm_models()
    return jsonify({"models": models})


@llm_bp.route("/models", methods=["POST"])
def create_model():
    """Add a new LLM model."""
    data = request.json

    # Validate required fields
    required_fields = ["name", "modelId", "apiKey"]
    for field in required_fields:
        if field not in data:
            return jsonify({"error": f"Missing required field: {field}"}), 400
        if not data[field] or not isinstance(data[field], str):
            return jsonify({"error": f"Invalid value for field: {field}"}), 400

    model = LLMModel(
        id=data["modelId"],
        name=data["name"],
        apiKey=data["apiKey"],
        apiEndpoint=data.get("apiEndpoint", None),
    )
    db_utils.add_llm_model(model)
    return jsonify({"message": "Model added successfully"}), 201


@llm_bp.route("/models/<path:model_id>", methods=["DELETE"])
def remove_model(model_id):
    if not model_id:
        return jsonify({"error": "Model ID is required"}), 400

    decoded_model_id = urllib.parse.unquote(model_id)
    success = db_utils.delete_llm_model(decoded_model_id)

    if success:
        return jsonify({"message": "Model deleted successfully"}), 200
    else:
        return jsonify({"error": "Model not found"}), 404


@llm_bp.route("/test", methods=["POST"])
def test_model():
    """Test an LLM model with a prompt."""
    data = request.json
    skip_lookup = request.args.get("skipLookup", "").lower() == "true"

    if not data or not isinstance(data, dict):
        return jsonify({"error": "Invalid request body"}), 400

    required_fields = ["modelId", "prompt"]
    for field in required_fields:
        if field not in data:
            return jsonify({"error": f"Missing required field: {field}"}), 400
        if not data[field] or not isinstance(data[field], str):
            return jsonify({"error": f"Invalid value for field: {field}"}), 400

    prompt = data["prompt"]
    try:
        if skip_lookup:
            model = LLMModel(
                id=data["modelId"],
                name=data.get("name", None) or data["modelId"],
                apiKey=data.get("apiKey", None),
                apiEndpoint=data.get("apiEndpoint", None),
            )
        else:
            model = llm_utils.get_model_by_id(data["modelId"])
            if not model:
                return jsonify({"error": "Model not found"}), 404
        response = llm_utils.test_prompt(model, prompt)
        return jsonify({"response": response}), 200
    except Exception as e:
        return jsonify({"error": f"{str(e)}"}), 500

File: server/go/routes/llm.go
package routes

import (
	"log"
	"net/url"
	"redefine/server/db"
	"redefine/server/llm"
	_ "redefine/server/llm/providers"
	"redefine/server/types"
	"strconv"

	"github.com/gin-gonic/gin"
)

func setupLLMRoutes(api *gin.RouterGroup) {
	llmGroup := api.Group("/llm")

	llmGroup.GET("/models", getModels)

	llmGroup.POST("/models", createModel)

	llmGroup.DELETE("/models/*model_id", removeModel)

	llmGroup.POST("/test", testModel)
}

func getModels(c *gin.Context) {
	llmModels, err := db.GetLLMModels()
	if err != nil {
		log.Printf("Error getting LLM models: %v", err)
		c.JSON(500, gin.H{"error": "Failed to retrieve LLM models"})
		return
	}

	sanitized := make([]types.LLMModel, len(llmModels))
	for i, model := range llmModels {
		sanitized[i] = model
		sanitized[i].APIKey = apiKeyPreview(model.APIKey)
	}

	c.JSON(200, gin.H{"models": sanitized})
}

func createModel(c *gin.Context) {
	var request struct {
		Name        string `json:"name"`
		ModelId     string `json:"modelId"`
		ApiKey      string `json:"apiKey"`
		ApiEndpoint string `json:"apiEndpoint"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		c.JSON(400, gin.H{"error": "Invalid request data"})
		return
	}

	if request.ModelId == "" || request.Name == "" || request.ApiKey == "" {
		c.JSON(400, gin.H{"error": "Missing required fields"})
		return
	}

	model := types.LLMModel{
		ID:          request.ModelId,
		Name:        request.Name,
		APIKey:      request.ApiKey,
		APIEndpoint: request.ApiEndpoint,
	}

	_, err := db.AddLLMModel(model)
	if err != nil {
		log.Printf("Error adding LLM model: %v", err)
		c.JSON(500, gin.H{"error": "Failed to create LLM model"})
		return
	}

	c.JSON(201, gin.H{"message": "Model added successfully"})
}

func removeModel(c *gin.Context) {

	modelID := c.Param("model_id")
	if modelID == "" {
		c.JSON(400, gin.H{"error": "Model ID is required"})
		return
	}

	decodedModelID, err := url.QueryUnescape(modelID)
	if err != nil {
		c.JSON(400, gin.H{"error": "Invalid model ID"})
		return
	}

	if len(decodedModelID) > 0 && decodedModelID[0] == '/' {
		decodedModelID = decodedModelID[1:]
	}

	log.Printf("Deleting LLM model with ID: %s", decodedModelID)

	success, err := db.DeleteLLMModel(decodedModelID)
	if err != nil {
		log.Printf("Error deleting LLM model: %v", err)
		c.JSON(500, gin.H{"error": "Failed to delete LLM model"})
		return
	}

	if !success {
		c.JSON(404, gin.H{"error": "Model not found"})
		return
	}

	c.JSON(200, gin.H{"message": "Model deleted successfully"})
}

func testModel(c *gin.Context) {
	var request struct {
		ModelId     string `json:"modelId"`
		Prompt      string `json:"prompt"`
		ApiKey      string `json:"apiKey"`
		ApiEndpoint string `json:"apiEndpoint"`
		Name        string `json:"name"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		c.JSON(400, gin.H{"error": "Invalid request data"})
		return
	}

	if request.ModelId == "" || request.Prompt == "" {
		c.JSON(400, gin.H{"error": "Missing required fields"})
		return
	}

	skipLookup, _ := strconv.ParseBool(c.Query("skipLookup"))

	var model *types.LLMModel
	var response string
	var err error
	if skipLookup {

		model = &types.LLMModel{
			ID:          request.ModelId,
			Name:        request.Name,
			APIKey:      request.ApiKey,
			APIEndpoint: request.ApiEndpoint,
		}

		if model.Name == "" {
			model.Name = request.ModelId
		}
	} else {

		model, err = db.GetLLMModelByID(request.ModelId)
		if err != nil {
			log.Printf("Error getting LLM model: %v", err)
			c.JSON(500, gin.H{"error": "Failed to retrieve LLM model"})
			return
		}

		if model == nil {
			c.JSON(404, gin.H{"error": "Model not found"})
			return
		}
	}

	response, err = llm.TestPrompt(model, request.Prompt, !skipLookup)

	if err != nil {
		log.Printf("Error testing LLM model: %v", err)
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.JSON(200, gin.H{"response": response})
}

func apiKeyPreview(key string) string {
	if key == "" {
		return ""
	}

	runes := []rune(key)
	if len(runes) <= 4 {
		return key
	}

	return string(runes[len(runes)-4:])
}
